### ISSUE TITLE
Parsing Content-Disposition header

### ISSUE THREAD

**Issue by whitelessk** [zbateson/mail-mime-parser#80]
I have an email with such attachment (repeated 😁 =F0=9F=98=81 symbols):

```
Content-Disposition: attachment; 
	filename*0="=?UTF-8?Q?*_=F0=9F=98=81=F0=9F=98=81=F0?=
 =?UTF-8?Q?=9F=98=8";
	filename*1="1=F0=9F=98=81=F0?=
 =?UTF-8?Q?=9F=98=81"; filename*2="=F0=9F=98=81.docx?="

```
For now the filename is parsed incorrectly:
* 😁😁���=8";    filename*1="1😁😁"; filename*2="😁.doc

But filename should look like * 😁😁😁😁😁😁.docx

It would be nice to concatenate filename (in case Parameter Value Continuations as described by RFC2184) before decoding to UTF-8.

**Comment by zbateson** [comment]
Hi @whitelessk --

Parameter Value Continuations have been supported since 1.0.  Current release is 1.1.

Having said that, your example looks wrong and probably wouldn't work anyway.  It should look more like this:

``` filename*0=utf-8''%F0%9F%98 ```, etc...

In your example there's no '*' after the parameter name, and it's combining a bit of RFC2184/2231 (2231 supercedes 2184) with 2047 for some reason.

Is this an actual example from a real mail client?  Or where did you find these headers? :)

**Comment by whitelessk** [comment]
Yes. This is an actual example, but I don't know source mail client (incoming email). Sorry for formating issue (* was stripped), fixed it. Tried to open such email in Thunderbird and attachment filename parsed correctly. Thats why I am interesting in this question.

**Comment by zbateson** [comment]
Okay, thanks for updating that -- so now that's an exact copy of how the header's formatted?  I'm really surprised that's allowed to be honest!

Just to confirm, you're on a 1.x version of the library?  You should at least get just one "filename" parameter accessible, though I'm pretty sure it won't give you a valid value I do think some parts, like the first, will be converted.

**Comment by zbateson** [comment]
Hmmm, still looks weird, are you sure you copy/pasted those headers?

Note that the RFC2231 spec needs two ``` ' ``` characters following the name, one's supposes to be for character encoding, the other language, if specified.  In your example the contents are quoted parts ``` " ```.  So yours is ``` filename*0="..."; ``` which is weird that it's "contained" within double quotes as well, versus the spec: ``` filename*0=''...; (no ending single quotes).

**Comment by whitelessk** [comment]
I am using 1.1 version. Headers copied as is. I cleaned private data and posted full email here https://pastebin.com/M25njJRi
Tried the following code:
```
$parts = $mail->getAllAttachmentParts();
        foreach ($parts as $part) {
            if ($part instanceof MessagePart) {
                $file_name = $part->getFilename();
                echo $file_name . PHP_EOL;
            }
        }
```
Everything works fine except few emails.

**Comment by zbateson** [comment]
Wow, and it was generated in 2018 unless you changed that.  I found a relevant mozilla-mail bug report from 16 years ago and was hoping your email was just old or something: https://bugzilla.mozilla.org/show_bug.cgi?id=193439

Anyways, it also has a few more examples of what other mail clients were producing, at least back then that I can hopefully incorporate into a fix.

**Comment by whitelessk** [comment]
Yes, the email is from 2018. I have read this article https://blog.nodemailer.com/2017/01/27/the-mess-that-is-attachment-filenames/ and understood that some Chinese clients (QQ webmail) just do not follow the standards.
Also, I have found a mixed Content-Type example in https://tools.ietf.org/html/rfc2231#section-4.1

```
Content-Type: application/x-stuff
    title*1*=us-ascii'en'This%20is%20even%20more%20
    title*2*=%2A%2A%2Afun%2A%2A%2A%20
    title*3="isn't it!"
```
Maybe it would be helpful.

**Comment by zbateson** [comment]
Fortunately all those work -- it seems the spec allows double-quoted parts as you noticed, and I programmed that in fine and have tests for it.  Also the bad examples in the blog post all work fine as well it seems.

The only one not working is the originally reported one.  It's a simple fix -- basically if I code it so RFC 2047 (=?....?=) encoded parts _to not allow_ double quotes in them, it fixes it.  Technically speaking, any printable character is allowed in RFC2047 though... so if I go further, and say double quotes are not allowed in an RFC2047 encoded part, when used within a parameter's value, which is specifically not permitted by the spec, I am hopefully fine.

To clarify, this would still be fine:

Subject: =?us-ascii?q?This_is_"okay"?=

This will not work any longer by my proposed fix:

Content-Type: text/plain; filename==?us-ascii?q?no_longer_"okay"?=

It is possible that some implementation somewhere will create something like that, I can't say for certain... but as far as allowing things not permitted by the spec goes, I think this is reasonable for now... if/when we run into something that actually has double quotes in a UFC2047 part, and as a parameter's value, we can deal with it then... what do you think?

**Comment by whitelessk** [comment]
I think that this is the right solution at the moment. I am ready to test the fix on my email array.

**Comment by zbateson** [comment]
Just pushed to master, have a go and let me know please.

**Comment by whitelessk** [comment]
@zbateson, I have tested changes and everything works perfectly. Filename header parsed correctly without new bugs. Thank you!

**Comment by zbateson** [comment]
Perfect, thanks for testing @whitelessk -- released in 1.1.1.
